File: generators/Class.ts
import { generateFromJsonMethod } from "./fromJson";
import { generateInsertMethod } from "./insertMethod";
import { generateStaticColumnNames } from "./staticColumnNames";
import { generateUpdateMethod } from "./updateMethod";
import { DartClass, Definitions } from "./types";
import { getDartTypeByFormat, snakeCasingToPascaleCasing } from "./utils";

export const generateDartClasses = (definitions: Definitions): DartClass[] => {
  const generatedClasses: DartClass[] = [];

  // Object entries
  for (const [tableName, table] of Object.entries(definitions)) {
    let dartCode = "";
    const { properties, required } = table;
    const className = snakeCasingToPascaleCasing(tableName);

    // Class definition
    dartCode += `class ${className} implements SupadartClass<${className}> {\n`;

    // Attributes
    for (const propertyName in properties) {
      const property = properties[propertyName];
      const dartType = getDartTypeByFormat(property.format);

      // Add question mark for optional fields (not in "required")
      const isOptional = !table.required.includes(propertyName);
      dartCode += `final ${dartType}${
        isOptional ? "?" : ""
      } ${propertyName};\n`;
    }

    // Constructor
    dartCode += `\n const ${className}({\n`;
    for (const propertyName in properties) {
      const isRequired = required.includes(propertyName);
      dartCode += `${
        isRequired ? "required this." : "this."
      }${propertyName},\n`;
    }
    dartCode += `});\n\n`;

    // Table name
    dartCode += `static String get table_name => '${tableName}';\n`;

    // Static column names
    dartCode += generateStaticColumnNames(properties);

    // Helper functions
    dartCode += generateInsertMethod(properties, required);
    dartCode += generateUpdateMethod(properties);
    dartCode += generateFromJsonMethod(className, properties, required);

    dartCode += "}\n\n";
    generatedClasses.push({
      className,
      classCode: dartCode,
    });
  }

  return generatedClasses;
};

File: cli/lib/generator/class.dart
import 'package:yaml/yaml.dart';

import 'converter_methods.dart';
import 'dart_class.dart';
import 'format_to_dart_type.dart';
import 'from_json_method.dart';
import 'insert_method.dart';
import 'static_column_names.dart';
import 'swagger.dart';
import 'update_method.dart';
import 'utils.dart';

List<DartClass> generateDartClasses(
    DatabaseSwagger swagger, YamlMap? mappings) {
  List<DartClass> generatedClasses = [];
  // Object entries
  swagger.definitions.forEach((tableName, table) {
    String dartCode = "";

    final columns = table.columns;
    final requiredFields = table.requiredFields;

    final className = tableNameToClassName(tableName, mappings);

    // Class definition
    dartCode += 'class $className implements SupadartClass<$className> {\n';

    // Attributes
    columns.forEach((columnName, columnDetails) {
      final dartType = postgresFormatToDartType(columnDetails.postgresFormat);

      // Add question mark for optional fields (not in "required")
      final isOptional = !requiredFields.contains(columnName);
      dartCode +=
          'final ${dartType.type}${isOptional ? "?" : ""} $columnName;\n';
    });

    // Constructor
    dartCode += '\n const $className({\n';
    columns.forEach((propertyName, _) {
      final isRequired = requiredFields.contains(propertyName);
      dartCode += '${isRequired ? "required this." : "this."}$propertyName,\n';
    });
    dartCode += '});\n\n';

    // Table name
    dartCode += 'static String get table_name => \'$tableName\';\n';

    // Static column names
    dartCode += generateStaticColumnNames(columns);

    // Methods
    dartCode += generateConverterMethod(className);
    dartCode += generateConverterSingleMethod(className);
    dartCode += generateInsertMethod(columns, requiredFields);
    dartCode += generateUpdateMethod(columns);
    dartCode += generateFromJsonMethod(className, columns, requiredFields);

    dartCode += '}\n\n';
    generatedClasses.add(
      DartClass(
        className: className,
        classCode: dartCode,
      ),
    );
  });

  return generatedClasses;
}
